packages/next/src/build/webpack/loaders/next-app-loader.ts

```diff
@@ -634,20 +639,30 @@ const nextAppLoader: AppLoader = async function nextAppLoader() {
   // This can be more efficient than checking them with `fs.stat` one by one
   // because all the thousands of files are likely in a few possible directories.
   // Note that it should only be cached for this compilation, not globally.
-  const filesInDir = new Map<string, Set<string>>()
   const fileExistsInDirectory = async (dirname: string, fileName: string) => {
-    const existingFiles = filesInDir.get(dirname)
-    if (existingFiles) {
-      return existingFiles.has(fileName)
+    // I don't think we should ever hit this code path, but if we do we should handle it gracefully.
+    if (this._compilation === undefined) {
+      try {
+        return (await getFilesInDir(dirname).catch(() => new Set())).has(
+          fileName
+        )
+      } catch (e) {
+        return false
+      }
     }
-    try {
-      const files = await getFilesInDir(dirname)
-      const fileNames = new Set<string>(files)
-      filesInDir.set(dirname, fileNames)
-      return fileNames.has(fileName)
-    } catch (err) {
-      return false
+    const map =
+      filesInDirMapMap.get(this._compilation) ||
```

Comment:
Why do we need to store in a map with the compilation as key, couldn't we hoist the map to module scope and re-use that way?

Reply:
Yeah so the original comment mentioned being cached per compilation. We'd need some way of invalidating if we scoped it to the module itself.

Reply:
Where is this invalidating on HMR?

Reply:
I assume that's a different compilation, so it'd have a different cache stored in the WeakMap. That's why I didn't use a straight map.

Reply:
Would be good to double check although considering our HMR tests are fine this sounds likely.